File: infer/tests/codetoanalyze/cpp/impurity/global_test.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
static int x;
int a[3];

void modify_global_primitive_impure() { x = 10; }

void modify_global_array_impure() { a[0] = 0; }

void call_modify_global() {
  modify_global_primitive_impure();
  modify_global_array_impure();
}

// modifies local static arr
int* local_static_var_impure() {

  // Lifetime of a static variable is throughout the program.
  static int arr[100];

  /* Some operations on arr[] */
  arr[0] = 10;
  arr[1] = 20;

  return arr;
}

void modify_global_inside_lamda_impure() {
  auto lam = [&]() { x++; };
}

File: infer/tests/codetoanalyze/java/impurity/GlobalTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class GlobalTest {
  public static int s = 0;
  public static Foo foo;

  class Foo {

    int x = 0;
    // modifies global var 's' hence impure
    void set_impure() {
      s = 10;
    }
  }

  void incr(Foo foo, int i) {
    foo.x += i;
  }

  // calls foo which modifies global var
  void call_set_impure() {
    Foo f = new Foo();
    f.set_impure();
  }

  // foo is global which is modified by incr.
  void global_mod_via_argument_passing_impure(int size, Foo f) {
    for (int i = 0; i < size; i++) {
      incr(foo, i);
    }
  }

  // aliased_foo is aliasing a global and then is modified by incr.
  void global_mod_via_argument_passing_impure_aliased(int size, Foo f) {
    Foo aliased_foo = foo; // Inferbo can't recognize aliasing here
    // and assumes aliased_foo is in [-oo,+oo] not in foo
    for (int i = 0; i < size; i++) {
      incr(aliased_foo, i);
    }
  }
}
